internal/perms/output_fields.go

```diff
@@ -6,68 +6,105 @@ import (
 	"github.com/hashicorp/boundary/globals"
 )
 
-// OutputFieldsMap is used to store information about allowed output fields in
+// OutputFields is used to store information about allowed output fields in
 // grants
-type OutputFieldsMap map[string]bool
+type OutputFields struct {
+	fields map[string]bool
+}
 
-// AddFields adds the given fields and returns the map.
-func (o OutputFieldsMap) AddFields(input []string) (ret OutputFieldsMap) {
+// AddFields adds the given fields and returns the interface. It is safe to call
+// this on a nil object, which will create a new object and add the fields to
+// it; if relying on this make sure to assign to the output, e.g.:
+//
+//	outFields = outFields.AddFields([]string{"foo", "bar"})
+//
+// Notes:
+//
+// - Adding non-nil but empty input will be construed as "no fields"
+//
+// - Fields compose, they do not overwrite; if you want to start over, create a
+// new OutputFields struct
+func (o *OutputFields) AddFields(input []string) *OutputFields {
+	ret := o
+	if ret == nil {
+		ret = new(OutputFields)
+	}
 	switch {
+	case input == nil:
+		// Do nothing
+
 	case len(input) == 0:
-		if o == nil {
-			return o
+		// Ensure we set to non-nil if it isn't already to capture that input is
+		// not empty but fields are not being added
+		if ret.fields == nil {
+			ret.fields = make(map[string]bool)
 		}
-		return o
-	case o == nil:
-		ret = make(OutputFieldsMap, len(input))
-	case len(o) == 1 && o["*"]:
-		return o
+
+	case len(ret.fields) == 1 && ret.fields["*"]:
+		// Again do nothing, there's nothing to add
+
 	default:
-		ret = o
-	}
-	for _, k := range input {
-		if k == "*" {
-			ret = OutputFieldsMap{k: true}
-			return
+		// Ensure the map is valid
+		if ret.fields == nil {
+			ret.fields = make(map[string]bool, len(input))
+		}
+
+		// Go through and add fields
+		for _, k := range input {
+			if k == "*" {
+				ret.fields = map[string]bool{k: true}
+				return ret
+			}
+			ret.fields[k] = true
 		}
-		ret[k] = true
 	}
-	return
-}
 
-func (o OutputFieldsMap) HasAll() bool {
-	return o["*"]
+	return ret
 }
 
-// Fields returns an alphabetical string slice of the fields in the map
-func (o OutputFieldsMap) Fields() (ret []string) {
-	if o == nil {
+// Fields returns an alphabetical string slice of the fields in the map. The
+// return value will be nil if fields are unset (e.g. we'd use the defaults in
+// SelfOrDefaults), and non-nil but empty if no fields are allowed. It is safe
+// to call this on a nil object; it will return a nil slice.
+func (o *OutputFields) Fields() (ret []string) {
+	if o == nil || o.fields == nil {
```

Comment:
I can get behind using the difference between a `nil` map and an empty slice internally, but should we really expose this as our interface? It's almost never an important distinction in Go. I'd like this to be more explicit, maybe
```go
func (o *OutputFields) Fields() (bool, []string) {
```
and have the `bool` be `false` if the fields are unset?

Reply:
Sure.